File: app/src/main/java/com/riders/thelab/ui/base/BaseActivity.java
package com.riders.thelab.ui.base;

import android.os.Bundle;

import androidx.annotation.Nullable;

import javax.inject.Inject;

import dagger.android.support.DaggerAppCompatActivity;

public class BaseActivity<V extends BaseView> extends DaggerAppCompatActivity {

    @Inject
    public V view;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        view.onCreate();
    }



    @Override
    protected void onDestroy() {
        view.onDestroy();
        super.onDestroy();
    }
}

File: app/src/main/java/com/riders/thelab/ui/base/BaseActivity.kt
package com.riders.thelab.ui.base

import android.os.Bundle
import androidx.activity.ComponentActivity
import com.riders.thelab.core.bus.Listen
import timber.log.Timber
import java.lang.reflect.InvocationTargetException

open class BaseActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.i("onCreate()")
        subscribeToKotlinBus()
    }

    private fun subscribeToKotlinBus() {
        Timber.i("subscribeToKotlinBus()")
        javaClass.declaredMethods
            .filter { it.isAnnotationPresent(Listen::class.java) }
            .forEach {
                try {
                    it.invoke(this)
                } catch (e: IllegalAccessException) {
                    // Utils.logException(e)
                    Timber.e(e)
                } catch (e: InvocationTargetException) {
                    // Utils.logException(e)
                    Timber.e(e)
                }
            }
    }
}